### DISCUSSION TITLE
[Bug]: 7.15.0 causes an issue with FieldErrors keys

### DISCUSSION THREAD

**Discussion by djMax** [react-hook-form/react-hook-form#6476] [Q&A]
### Version Number

7.15.0

### Codesandbox/Expo snack

https://codesandbox.io/s/reverent-silence-piwiw?file=/src/Input.tsx

### Steps to reproduce

I'm afraid this is a complicated Typescript dance. We have a custom Input component that tries to make the process of using react-hook-form a little easier. It is defined as follows:

```
export interface GbUseFormExtensions<T extends Record<string, any>>
  extends ReturnType<typeof useForm> {
  setters: { [P in keyof T]: (value: any) => void };
  triggers: { [P in keyof T]: (value: any) => void };
  triggerAll: () => void;
}

export function useFormWithSchema<T extends object>(
  schema: T,
  formOptions: Parameters<typeof useForm>[0] = {},
): GbUseFormExtensions<T> {
/* IMPL HERE */
}
```

This has worked fine for some time up through 7.14.2. When updating to 7.15.0 today, I get this error:

```
src/Input/index.tsx(106,11): error TS2536: Type 'Extract<keyof T, string>' cannot be used to index type 'FieldError'.
src/Input/index.tsx(109,24): error TS2536: Type 'Extract<keyof T, string>' cannot be used to index type 'FieldError'.
```

Those lines are:
```
      if (formHook.formState.errors?.[name]) {
```
and
```
        errorMessage = formHook.formState.errors[name]?.message;
```

The input name property is declared as part of the Input component and meant to give the name of the field in the form schema:

```
export interface GbInputProps<T> extends Omit<InputProps, 'ref'> {
  name?: Extract<keyof T, string>;
  formHook?: GbUseFormExtensions<T>;
```

(T will be inferred from the passed in formHook from the useFormWithSchema hook above.)

I get that this is complex, but hoping maybe something about the change made to 7.15 will make it obvious to someone who isn't me.

### Expected behaviour

Working compile.

### What browsers are you seeing the problem on?

Chrome

### Relevant log output

```shell
src/Input/index.tsx(106,11): error TS2536: Type 'Extract<keyof T, string>' cannot be used to index type 'FieldError'.
src/Input/index.tsx(109,24): error TS2536: Type 'Extract<keyof T, string>' cannot be used to index type 'FieldError'.
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
This is mostly due to the TS improvement made in 7.15.0

**Comment by djMax** [comment]
I figured, I'm just hoping someone says "oh yeah, that's because we changed the string type to IFoo, so if you change your keyof T to MagicBar<T> it will all be fine!" :)

**Comment by kotarella1110** [comment]
The type error also seems to occur in `7.14.2`.

**Comment by kotarella1110** [comment]
The reason for the type error is that the `name` prop is optional.

```diff
export interface GbInputProps<T> extends Omit<InputProps, "ref"> {
- name?: Extract<keyof T, string>;
+ name: Extract<keyof T, string>;
  formHook?: GbUseFormExtensions<T>;
  floatingLabel?: string;
  outlined?: boolean;
  watch?: boolean;
  formValueFormatter?: (value: any) => string;
}
```

**Comment by bluebill1049** [comment]
thanks very much for the investigation @kotarella1110 I think this one should be convert to Q&A instead.

**Comment by djMax** [comment]
The snack is fixed, but the real code is not. What's very confusing to me is the error: 

```
src/Input/index.tsx(106,11): error TS2536: Type 'Extract<keyof T, string>' cannot be used to index type 'FieldError'.
```

It's not FieldError, right? It should be FieldErrors. The code looks right, but it really seems like it's confusing Typescript in some way.
